### DISCUSSION TITLE
Creating good parser from templates

### DISCUSSION THREAD

**Discussion by kilomeow** [TiddlyWiki/TiddlyWiki5#5541] [Q&A]
In my widget code I'm trying to render some tiddler using template:

```
var parser = $tw.wiki.parseTiddler("$:/my/template"),
  widgetNode = $tw.wiki.makeWidget(parser,
        { variables: {currentTiddler: "$:/my/tiddler"}}),
  container = $tw.fakeDocument.createElement("div");

widgetNode.render(container,null);
```
`$:/my/template` starts with pragma `\rules only filteredtranscludeinline transcludeinline macrodef macrocallinline html` so it's parsing correctly and it works. However, in a result I get a lot of `<p>` tags from template code. 
I tried to use `{parseAsInline: true}` as an argument to `parseTiddler` but for some reason it doesn't work in any way -- I'm still getting a lot of `<p>` tags which mess resulting html. How to build a good template parser?

**Comment by saqimtiaz** [comment]
It would be helpful if you share both what you are trying to accomplish, as well as the contents of your template.

**Reply by kilomeow** [reply]
template:
```
\rules only filteredtranscludeinline transcludeinline macrodef macrocallinline html

<html>
  <head>
    <title>
      Blog
    </title>
  </head>
<body>

<style>
  html {
      font-family: Arial, sans-serif;
      font-size: .938em;
      line-height: 1.4;
      background-color: #feffff;
  }

   /* more css here */
</style>

<div class="top-panel">
  <div class="logo">
  </div>
  <div class="about">
  <span class="blog-title">
    <h1>My blog</h1>
  </span>
  </div>
</div>

<div class="flow">
<$transclude />
</div>
</body>
</html>
```
For a start I just want to see the content of my tiddler instead of <$transclude /> in this template. A bit later I want to add transclusion of special fields of my tiddler for creating page title, custom icon etc, but for a start just this. 

The result html is:

```
<html>
  <head>
    <title>
      Fission powered blog
    </title>
  </head>
<body><p><style>
  html {
      font-family: Arial, sans-serif;
      font-size: .938em;
      line-height: 1.4;
      background-color: #feffff;
      
        /* more css here */
  }
</style></p><p><div class="top-panel">
  <div class="logo">
  </div>
  <div class="about">
  <span class="blog-title">
    <h1>Fission powered blog</h1>
  </span>
  </div>
</div></p><p><div class="flow">
<my-content></my-content>
</div>
</p></body>
</html>
```

You could see here a lot of annoying `<p>` tags

**Reply by saqimtiaz** [reply]
The p tags are due to the empty lines in your template. You can remove them or try replacing the \rules line with:

```
\rules only filteredtranscludeinline transcludeinline macrodef macrocallinline html whitespace
\whitespace trim

<html>
....
```

**Reply by kilomeow** [reply]
oh thank you so much !! whitespace rule rocks. however it leaves me with `<p>`s for every empty line (`\n\n`) and I want to fix it a bit later. It would be cool to have eye-pleasent template source, but now it's ok to have it without empty lines.

**Comment by Jermolene** [comment]
Hi @kilomeow by default the parser will assume the text is in block mode, which means that `<p>` tags are generated and block-level syntax like headings is recognised.

You can force inline mode as follows which will restrict the parser to inline parsing rules:

```js
var parser = $tw.wiki.parseTiddler("$:/my/template",{parseAsInline: true}),
  widgetNode = $tw.wiki.makeWidget(parser,
        { variables: {currentTiddler: "$:/my/tiddler"}}),
  container = $tw.fakeDocument.createElement("div");

widgetNode.render(container,null);
```
